### PR TITLE
Block gallery: delete image with delete or backspace keys

### LINKED ISSUE #14816: Block gallery: DELETE and BACKSPACE do not remove the image
When a gallery image is selected, pressing DELETE or BACKSPACE keys should delete it.

There is [some code](https://github.com/WordPress/gutenberg/blob/master/packages/block-library/src/gallery/gallery-image.js#L59) for this but doesn't work.

### PR DESCRIPTION
**Pull request by oandregal**
Fixes https://github.com/WordPress/gutenberg/issues/14816

Peek 2019-04-04 16-51

### CODE REVIEW

File: packages/block-library/src/gallery/gallery-image.js [hunk 1]
```diff
@@ -127,7 +128,7 @@ class GalleryImage extends Component {
 		// Disable reason: Each block can be selected by clicking on it and we should keep the same saved markup
 		/* eslint-disable jsx-a11y/no-noninteractive-element-interactions, jsx-a11y/onclick-has-role, jsx-a11y/click-events-have-key-events */
 		return (
-			<figure className={ className } tabIndex="-1" onKeyDown={ this.onKeyDown } ref={ this.bindContainer }>
+			<figure className={ className }>
```
Comment (gziolo) [inline packages/block-library/src/gallery/gallery-image.js]: Is code comment explaining disable reason still valid?

Comment (nosolosw) [reply packages/block-library/src/gallery/gallery-image.js]: Was not! Thank you. Fixed at be94811

**Review by youknowriad — APPROVED** [review]
the code changes are good, but I'd like an accessibility check here.

**Comment by afercia** [comment]
I've quickly tested with various browser / screen reader combinations and the removal works. A bit surprisingly because I wasn't expecting `onKeyDown` to work on non-natively interactive elements when screen readers are not in "focus mode". We've already faced this issue in other components, @aduth might remember something about it.

Regardless, it works.

A pre-existing issue (triggered also when removing an image via the dedicated button) is that focus is not managed. When the selected image (currently focused element) gets removed, there's no focused element any longer so a focus loss happens.

Focus should be moved to the most logical place. Not sure what the best place is 🙂 Ideally:
- to the previous image, if any
- otherwise to the following image, if any
- otherwise if it's the last image: to the block focusable container

I do realize this would be complicated so I'd be OK with always moving focus to the block focusable container: at least it would avoid a complete focus loss.

I also do realize it's not strictly related to this PR so it might be addressed separately but since we're here... 🙂

File: packages/block-library/src/gallery/gallery-image.js [hunk 2]
```diff
@@ -18,9 +18,9 @@ class GalleryImage extends Component {
 	constructor() {
 		super( ...arguments );
 
-		this.onImageClick = this.onImageClick.bind( this );
+		this.onSelectImage = this.onSelectImage.bind( this );
 		this.onSelectCaption = this.onSelectCaption.bind( this );
-		this.onKeyDown = this.onKeyDown.bind( this );
+		this.onRemoveImage = this.onRemoveImage.bind( this );
 		this.bindContainer = this.bindContainer.bind( this );
```
Comment (aduth) [inline packages/block-library/src/gallery/gallery-image.js]: `bindContainer` is unused and should be removed.

**Comment by aduth** [comment]
> A bit surprisingly because I wasn't expecting `onKeyDown` to work on non-natively interactive elements when screen readers are not in "focus mode". We've already faced this issue in other components, @aduth might remember something about it.

If I'm recalling correctly, there was a specific impact of both the specific key pressed and the effective `role` of the element in how keyboard events were treated between focus and browse modes (#5595).

**Comment by afercia** [comment]
> If I'm recalling correctly, there was a specific impact of both the specific key pressed and the effective role of the element in how keyboard events were treated between focus and browse modes 

Yep, evidently it works on images (implicit role=img).

**Comment by nosolosw** [comment]
Created an issue for the focus loss at https://github.com/WordPress/gutenberg/issues/14946

File: packages/block-library/src/gallery/gallery-image.js (hunk 2 above)
Comment (nosolosw) [reply packages/block-library/src/gallery/gallery-image.js]: mmm, notice that it's used by the `` to store a reference. Then that reference is used in `onRemoveImage`.

**Comment by nosolosw** [comment]
OK, I'm going to go ahead here and land this so it's in the list of bugfixes to port to WordPress 5.2.

Shameless self-promotion in case anyone has time to review related gallery bugfixes that deal with keyboard navigation and focus management: https://github.com/WordPress/gutenberg/pull/14930 and https://github.com/WordPress/gutenberg/pull/14821 I'm going to be out next week, but I'll try to save a couple of hours tomorrow or perhaps Sunday in case reviews come today when I' already AFK. It'll be nice to get those bugfixes in as well!

**Pull request merged**